src/PJ_axisswap.c

```diff
@@ -0,0 +1,227 @@
+    struct pj_opaque *Q = (struct pj_opaque *) P->opaque;
+    unsigned int i;
+    PJ_COORD in, out;
+
+    out = proj_coord_error();
+    in = obs.coo;
+
+    for (i=0; i<4; i++)
+        out.v[i] = in.v[Q->axis[i]] * Q->sign[i];
+
+    obs.coo = out;
+    return obs;
+}
+
+
+static PJ_OBS reverse_obs(PJ_OBS obs, PJ *P) {
+    struct pj_opaque *Q = (struct pj_opaque *) P->opaque;
+    unsigned int i;
+    PJ_COORD in, out;
+
+    out = proj_coord_error();
+    in = obs.coo;
+
+    for (i=0; i<4; i++) {
+        out.v[Q->axis[i]] = in.v[i] * Q->sign[i];
+    }
+
+    obs.coo = out;
+    return obs;
+}
+
+
+/***********************************************************************/
+PJ *PROJECTION(axisswap) {
+/***********************************************************************/
+    struct pj_opaque *Q = pj_calloc (1, sizeof (struct pj_opaque));
+    char *order;
+    size_t i;
+    int sign=1, i_axis=0;
+    int axes_count[4] = {0, 0, 0, 0};
+
+    if (0==Q)
+        return pj_default_destructor (P, ENOMEM);
+    P->opaque = (void *) Q;
+
+    /* read axis order */
+    if (pj_param (P->ctx, P->params, "torder").i) {
+        order = pj_param(P->ctx, P->params, "sorder").s;
+    } else {
+        return pj_default_destructor(P, PJD_ERR_MISSING_ARGS);
+    }
+
+    /* check that all characteds are valid */
+    for (i=0; i<strlen(order); i++)
+        if (strchr("1234-", order[i]) == 0) {
+            proj_log_error(P, "axisswap: unknown axis '%c'", order[i]);
+            return pj_default_destructor(P, PJD_ERR_AXIS);
+        }
+
+    /* no trainling minuses */
+    if (order[i-1] == '-') {
+        proj_log_error(P, "axisswap: trailing '-' not allowed in axis order specification");
+        return pj_default_destructor(P, PJD_ERR_AXIS);
+    }
+
+    /* count occurrences of each axis */
+    for (i=0; i<strlen(order); i++) {
+        if (order[i] == '-')
+            continue;
+        axes_count[ order[i] - '0' - 1 ]++;
+    }
+
+    /* check if axes are specified correctly */
+    for (i=0; i<4; i++) {
+        if (axes_count[i] > 1) {
+            proj_log_error(P, "axisswap: axis '%c' specified more than once", order[i]);
+            return pj_default_destructor(P, PJD_ERR_AXIS);
+        }
+        if (axes_count[i] < 1) {
+            proj_log_error(P, "axisswap: axis '%c' not specified", order[i]);
+            return pj_default_destructor(P, PJD_ERR_AXIS);
+        }
+    }
+
+    /* read axis order and sign */
+    for (i=0; i<strlen(order); i++) {
+        if (order[i] == '-') {
+            if (sign == -1)
+                return pj_default_destructor(P, PJD_ERR_AXIS);
+            sign = -1;
+            continue;
+        }
+        Q->axis[i_axis] = order[i] - '0' - 1;
+        Q->sign[i_axis] = sign;
+        i_axis++;
+        sign = 1;
+    }
+
+    P->fwdobs = forward_obs;
+    P->invobs = reverse_obs;
+    P->fwd3d  = forward_3d;
+    P->inv3d  = reverse_3d;
+    P->fwd    = forward_2d;
+    P->inv    = reverse_2d;
+
```

Comment:
Just before the code block above, it would make sense to add a check for whether the swap was valid for a given number of dimensions, and then using null pointers for the corresponding  fwd/inv combo if not.

Since pj_fwd, pj_inv, pj_fwd3d, pj_inv3d, pj_fwdobs and pj_invobs all check for nulls, and return appropriate error objects, you could eliminate the `if (Q->axis[i] < 3)` style checks, and get better run time performance, at the cost of a minimally longer setup time.

Reply:
Good suggestion. I'll include that.